lib/index.js

```diff
@@ -404,6 +566,157 @@ class AuthenticationPlugin {
+      throw new this.context.errors.BadRequestError('Missing request body');
+    }
+
+    const {
+      password,
+      token
+    } = request.input.body;
+
+    if (!password) {
+      throw new this.context.errors.BadRequestError('Missing "password" attribute');
+    }
+    if (typeof password !== 'string' || password.trim() === '') {
+      throw new this.context.errors.BadRequestError('Invalid password supplied. Must be a non-empty string');
+    }
+    if (!token) {
+      throw new this.context.errors.BadRequestError('Missing "token" attribute');
+    }
+    if (typeof token !== 'string') {
+      throw new this.context.errors.BadRequestError('Invalid token supplied');
+    }
+
+    let kuid;
+    try {
+      kuid = jsonwebtoken.verify(token, this.config.resetPasswordSecret).kuid;
+    }
+    catch (e) {
+      if (e instanceof jsonwebtoken.TokenExpiredError) {
+        this.context.errorsManager.throw('expired-token');
+      }
+
+      if (e instanceof jsonwebtoken.JsonWebTokenError) {
+        this.context.errorsManager.throw('invalid-token');
+      }
+
+      throw new this.context.errors.BadRequestError(e.message);
+    }
+
+    await this.validate(request, {password}, kuid, 'local', true);
+    await this.update(request, {password}, kuid);
+
+    const user = await this.getCredentialsFromUserId(kuid);
+
+    const loginRequest = new this.context.constructors.Request({
+      action: 'login',
+      body: {
+        password,
+        username: user._id
+      },
+      controller: 'auth',
+      strategy: 'local'
+    }, request.context);
+    const loginResponse = await this.context.accessors.execute(loginRequest);
+
+    return loginResponse.result;
+  }
+
+  // private
+  // --------------------------------------------------------------------------
+  _getResetPasswordToken(kuid) {
+    const options = {
+      issuer: 'kuzzle-plugin-auth-passport-local'
+    };
+
+    if (this.config.resetPasswordExpiresIn > -1) {
+      options.expiresIn = this.config.resetPasswordExpiresIn;
+    }
+
+    return jsonwebtoken.sign(
+      { kuid },
+      this.config.resetPasswordSecret,
+      options
+    );
+  }
+
+  /**
+   * @param {KuzzleRequest} request
+   * @param {{ username: string, password: string }} credentials
+   * @param {string} kuid
+   * @returns {Promise<boolean>}
+   */
+  async _validateCreate(request, credentials, kuid) {
+    if (!credentials.username) {
+      throw new this.context.errors.BadRequestError('Username required.');
+    }
+
+    if (!credentials.password) {
+      throw new this.context.errors.BadRequestError('Password required.');
+    }
+
+    const user = await this.getUsersRepository().get(credentials.username);
+
+    if (user !== null && kuid !== user.kuid) {
+      throw new this.context.errors.PreconditionError(`Login "${credentials.username}" is already used.`);
+    }
+
+    const newUser = new this.User();
+    newUser.kuid = kuid;
+
+    if (await newUser.getKuzzleUser() === null) {
```

Comment:
I don't understand how can this happen?

Reply:
You're right. I first thought is was somehow possible to call a strategy method directly from a plugin but it is not (except for privileged ones).
Kuzzle controllers ensure the user exists before any credentials can be set.

I will remove the condition.

Reply:
done